File: C++/가장먼노드.cpp
#include <string>
#include <vector>
#include<algorithm>
#include<queue>
using namespace std;
int visit[20001];
vector<int> node[20001];
int solution(int n, vector<vector<int>> edge) {
    int answer = 0,maxCnt=0;;
    for(int i=0;i<edge.size();i++){
        node[edge[i][0]].push_back(edge[i][1]);
        node[edge[i][1]].push_back(edge[i][0]);
    }
    for(int i=1;i<=n;i++){
        visit[i]=-1;
    }
    queue<int> q;
    q.push(1);
    visit[1]=0;
    while(!q.empty()){
        int cur = q.front();
        int curCnt = visit[cur];
        q.pop();
        for(int i=0;i<node[cur].size();i++){
            int next = node[cur][i];
            if(visit[next]==-1||curCnt+1<visit[next]){
                visit[next]=curCnt+1;
                maxCnt = max(maxCnt,visit[next]);
                q.push(next);
            }
        }
    }
    for(int i=1;i<=n;i++){
        if(visit[i]==maxCnt)answer++;
    }
    return answer;
}

File: java/가장먼노드.java
import java.util.*;
class Solution {
    ArrayList<ArrayList<Integer>> e = new ArrayList<>();
    int []dist ;
    public int solution(int n, int[][] edge) {
        dist = new int[n];
        for(int i=0;i<n;i++){
            dist[i]=0x3f3f3f3f;
        }
        for(int i=0;i<n;i++){
            e.add(new ArrayList<Integer>());
        }
        for(int i=0;i<edge.length;i++){
            int a = edge[i][0]-1;
            int b = edge[i][1]-1;
            e.get(a).add(b);
            e.get(b).add(a);
        }
        
        Queue<Integer> q = new LinkedList<>();
        q.add(0);
        dist[0]=0;
        int maxVal=0;
        while(!q.isEmpty()){
            int cur = q.poll();
            int now = dist[cur];
            for(int i=0;i<e.get(cur).size();i++){
                int next = e.get(cur).get(i);
                if(dist[next]>now+1){
                    dist[next]=now+1;
                    maxVal = Math.max(dist[next],maxVal);
                    q.add(next);
                }
            }
        }
        
        
        int answer = 0;
        for(int i=0;i<n;i++){
            if(dist[i]==maxVal)
                answer++;
        }
        return answer;
    }
}